packages/react-google-maps-api/src/components/drawing/Circle.tsx

```diff
@@ -96,71 +98,69 @@ export interface CircleProps {
   onUnmount?: (circle: google.maps.Circle) => void
 }
 
-export class Circle extends React.PureComponent<CircleProps, CircleState> {
-  static contextType = MapContext
-
-  registeredEvents: google.maps.MapsEventListener[] = []
-
-  state: CircleState = {
-    circle: null,
-  }
-
-  setCircleCallback = (): void => {
-    if (this.state.circle !== null && this.props.onLoad) {
-      this.props.onLoad(this.state.circle)
-    }
-  }
-
-  componentDidMount(): void {
-    const circle = new google.maps.Circle({
-      ...(this.props.options || {}),
-      map: this.context,
-    })
-
-    this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
-      updaterMap,
-      eventMap,
-      prevProps: {},
-      nextProps: this.props,
-      instance: circle,
-    })
-
-    this.setState(function setCircle() {
-      return {
-        circle,
-      }
-    }, this.setCircleCallback)
-  }
-
-  componentDidUpdate(prevProps: CircleProps): void {
-    if (this.state.circle !== null) {
-      unregisterEvents(this.registeredEvents)
-
-      this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
-        updaterMap,
-        eventMap,
-        prevProps,
-        nextProps: this.props,
-        instance: this.state.circle,
-      })
-    }
-  }
-
-  componentWillUnmount(): void {
-    if (this.state.circle !== null) {
-      if (this.props.onUnmount) {
-        this.props.onUnmount(this.state.circle)
+function Circle(props: CircleProps): JSX.Element {
+  const { options, onLoad, onUnmount } = props
+  const map = React.useContext(MapContext)
+  const prevProps: CircleProps = usePrevious<CircleProps>(props)
+
+  const [instance, setInstance] = React.useState<google.maps.Circle | null>(
+    null
+  )
+
+  React.useEffect(
+    function effect() {
+      if (map !== null) {
+        if (instance === null) {
```

Comment:
Maybe we can extract this also to a custom hook? It would need to receive a generic to indicate the instance type, but i think instance, setInstance and this useEffect can be in a generic useMapComponent hook